metamorphic: TestMeta failed [could not generate unique key]

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14071847?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14071847?buildTab=artifacts#meta) on refs/heads/master @ [5a2d7792b0cf](https://github.com/cockroachdb/pebble/commits/5a2d7792b0cf):


```
iter7.Next("qziovqcdah")
iter5.SeekLT("nnuvrwmu@20", "")
iter4.Last()
db1.Set("zfem@6", "djlvigsscy")
iter2.SeekPrefixGE("xjammurepm@11")
db1.RatchetFormatMajorVersion(013)
iter5.Prev("")
iter3.SeekLT("qkrncoailp@22", "")
iter6.SeekGE("rjkvc@17", "")
iter5.SeekGE("qkrncoailp@23", "")
db1.Set("tqjzsdrdnaqz@5", "")
iter7.SetBounds("xjammurepm@3", "xjammurepm@3")
iter5.Prev("")
iter4.Prev("")
iter3.First()
iter5.SeekLT("tqjzsdrdnaqz@5", "")
db1.Set("qwadvgbqocs@5", "fusatlvwhuyqro")
db1.IngestExternalFiles(external0, "tqjzsdrdnaqz@17", "tqjzsdrdnaqz@10", "", external0, "tqjzsdrdnaqz@10", "tqjzsdrdnaqz@3", "")

panic: could not generate unique key [recovered]
	panic: could not generate unique key [recovered]
	panic: could not generate unique key

goroutine 8 [running]:
testing.tRunner.func1.2({0xc8f220, 0xeba6e0})
	GOROOT/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x397
panic({0xc8f220?, 0xeba6e0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:208 +0xa5
panic({0xc8f220?, 0xeba6e0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).UniqueKeys(0xc0002361a0, 0x2, 0xc0001d90c0)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/key_generator.go:73 +0x245
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc0000f38c0)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:1333 +0x9a5
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:212 +0xd5e
github.com/cockroachdb/pebble/metamorphic.RunAndCompare(0xc000236000, {0x7ffd9b10b93b, 0xf}, {0xc0001dfb30, 0x3, 0x3})
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:199 +0x730
github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0xc000078760?, 0x0)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:94 +0x239
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x0?)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:54 +0x15
testing.tRunner(0xc000236000, 0xdf9078)
	GOROOT/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x3ad
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Reproduces with `TestParserRandom` too:

```
--- FAIL: TestParserRandom (0.15s)
    --- FAIL: TestParserRandom/config=default (0.15s)
panic: could not generate unique key [recovered]
	panic: could not generate unique key [recovered]
	panic: could not generate unique key

goroutine 1970 [running]:
testing.tRunner.func1.2({0xbb0740, 0xdd1980})
	/opt/hostedtoolcache/go/1.21.7/x64/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.21.7/x64/src/testing/testing.go:1548 +0x397
panic({0xbb0740?, 0xdd1980?})
	/opt/hostedtoolcache/go/1.21.7/x64/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	/home/runner/work/pebble/pebble/metamorphic/generator.go:208 +0xa5
panic({0xbb0740?, 0xdd1980?})
	/opt/hostedtoolcache/go/1.21.7/x64/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).UniqueKeys(0xc000a13040, 0x2, 0xc000137290)
	/home/runner/work/pebble/pebble/metamorphic/key_generator.go:73 +0x245
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc0000fab00)
	/home/runner/work/pebble/pebble/metamorphic/generator.go:1333 +0x9a5
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	/home/runner/work/pebble/pebble/metamorphic/generator.go:212 +0xd5e
github.com/cockroachdb/pebble/metamorphic.TestParserRandom.func1(0xc000a12ea0)
	/home/runner/work/pebble/pebble/metamorphic/parser_test.go:40 +0x308
testing.tRunner(0xc000a12ea0, 0xc0004a38c0)
	/opt/hostedtoolcache/go/1.21.7/x64/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1969
	/opt/hostedtoolcache/go/1.21.7/x64/src/testing/testing.go:1648 +0x3ad
```
https://github.com/cockroachdb/pebble/actions/runs/7978180827/job/21782738930

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -13,6 +13,7 @@ import (
 
 	"github.com/cockroachdb/pebble"
 	"github.com/cockroachdb/pebble/internal/randvar"
+	"github.com/cockroachdb/pebble/sstable"
 	"golang.org/x/exp/rand"
 )
 
@@ -1283,82 +1284,74 @@ func (g *generator) writerIngestExternalFiles() {
 	for i := range objs {
 		// We allow the same object to be selected multiple times.
 		id := g.externalObjects.rand(g.rng)
-
 		b := g.keyManager.objKeyMeta(id).bounds
-		// For now, set the bounds to the bounds of the entire object.
-		kr := pebble.KeyRange{
-			Start: b.smallest,
-			End:   b.largest,
-		}
+
+		objStart := b.smallest
+		objEnd := b.largest
 		if !b.largestExcl {
 			// Move up the end key a bit by appending a few letters to the prefix.
-			kr.End = append(g.prefix(kr.End), randBytes(g.rng, 1, 3)...)
+			objEnd = append(g.prefix(objEnd), randBytes(g.rng, 1, 3)...)
+		}
+		// Generate two random keys within the given bounds.
+		// First, generate a start key in the range [objStart, objEnd).
+		start := g.keyGenerator.RandKeyInRange(0.01, pebble.KeyRange{
+			Start: objStart,
+			End:   objEnd,
+		})
+		// Second, generate an end key in the range (start, objEnd]. To do this, we
+		// generate a key in the range [start, objEnd) and if we get `start`, we
+		// remap that to `objEnd`.
+		end := g.keyGenerator.RandKeyInRange(0.01, pebble.KeyRange{
+			Start: start,
+			End:   objEnd,
+		})
+		if g.cmp(start, end) == 0 {
+			end = objEnd
+		}
+
+		var syntheticSuffix sstable.SyntheticSuffix
+		if g.rng.Intn(2) == 0 {
+			// Try to set a synthetic suffix. We can only do so if the bounds don't
+			// have suffixes, so trim them.
+			start = g.prefix(start)
+			end = g.prefix(end)
+			if g.cmp(start, end) == 0 {
+				// Move the end bound up a bit.
+				end = append(g.prefix(end), randBytes(g.rng, 1, 3)...)
+			}
+			syntheticSuffix = g.keyGenerator.SkewedSuffix(0.1)
 		}
 
 		objs[i] = externalObjWithBounds{
 			externalObjID: id,
-			bounds:        kr,
+			bounds: pebble.KeyRange{
+				Start: start,
+				End:   end,
+			},
+			syntheticSuffix: syntheticSuffix,
 		}
 	}
 
-	// Sort the objects by their start bound.
-	sorted := make([]*externalObjWithBounds, numFiles)
-	for i := range sorted {
-		sorted[i] = &objs[i]
-	}
-	slices.SortFunc(sorted, func(a, b *externalObjWithBounds) int {
+	slices.SortFunc(objs, func(a, b externalObjWithBounds) int {
 		return g.cmp(a.bounds.Start, b.bounds.Start)
 	})
 
-	// We take the overall bounds of all the objects and generate 2*numFiles keys
-	// in that range to create the per-object bounds. In many cases, there won't
-	// be overlap between the object and its corresponding bounds but that's ok
-	// (we verify that when running the op).
-	overall := sorted[0].bounds
-	for i := 1; i < len(sorted); i++ {
-		if g.cmp(overall.End, sorted[i].bounds.End) < 0 {
-			overall.End = sorted[i].bounds.End
-		}
-	}
-	if g.cmp(overall.Start, overall.End) >= 0 {
-		panic(fmt.Sprintf("invalid bounds [%q, %q)", overall.Start, overall.End))
-	}
-	// If the bounds have the same prefix, we might not be able to generate enough unique keys.
-	if g.cmp(g.prefix(overall.Start), g.prefix(overall.End)) == 0 {
-		// Append some bytes to move up the end bound.
-		overall.End = append(g.prefix(overall.End), randBytes(g.rng, 1, 3)...)
-	}
-	// Generate 2*numFiles distinct keys and sort them. These will form the ingest
-	// bounds for each file.
-	keys := g.keyGenerator.UniqueKeys(2*numFiles, func() []byte {
-		return g.keyGenerator.RandKeyInRange(0.01, overall)
-	})
-
-	for i, o := range sorted {
-		if i > 0 && g.rng.Intn(4) == 0 {
-			// Sometimes use the previous end key as the start key; this will be common in practice.
-			o.bounds.Start = sorted[i-1].bounds.End
-		} else {
-			o.bounds.Start = keys[2*i]
-		}
-		o.bounds.End = keys[2*i+1]
-
-		if g.rng.Intn(2) == 0 {
-			// Try to set a synthetic suffix. We can only do so if the bounds don't
-			// have suffixes, so try to trim them.
-			start := g.prefix(o.bounds.Start)
-			end := g.prefix(o.bounds.End)
-			// If the trimmed bounds overlap with adjacent file bounds, we just don't
-			// set the suffix.
-			if g.cmp(start, end) < 0 &&
-				(i == 0 || g.cmp(sorted[i-1].bounds.End, start) <= 0) &&
-				(i == len(sorted)-1 || g.cmp(end, sorted[i+1].bounds.Start) <= 0) {
-				o.bounds.Start = start
-				o.bounds.End = end
-				o.syntheticSuffix = g.keyGenerator.SkewedSuffix(0.1)
+	// Trim bounds so that there is no overlap.
+	for i := 0; i < len(objs)-1; i++ {
+		if g.cmp(objs[i].bounds.End, objs[i+1].bounds.Start) > 0 {
+			objs[i].bounds.End = objs[i+1].bounds.Start
+			if objs[i].syntheticSuffix != nil {
+				// Bounds can't have suffixes when syntheticSuffix is used.
+				objs[i].bounds.End = g.prefix(objs[i].bounds.End)
 			}
 		}
 	}
+	// Some bounds might be empty now, remove those objects altogether. Note that
+	// the last object is unmodified, so at least that object will remain.
+	objs = slices.DeleteFunc(objs, func(o externalObjWithBounds) bool {
+		return g.cmp(o.bounds.Start, o.bounds.End) >= 0
+	})
+
 	// The batches we're ingesting may contain single delete tombstones that
 	// when applied to the writer result in nondeterminism in the deleted key.
 	// If that's the case, we can restore determinism by first deleting the keys
@@ -1378,6 +1371,11 @@ func (g *generator) writerIngestExternalFiles() {
 			}
 		}
 	}
+
+	// Shuffle the objects.
+	g.rng.Shuffle(len(objs), func(i, j int) {
+		objs[i], objs[j] = objs[j], objs[i]
+	})
 
 	g.add(&ingestExternalFilesOp{
 		dbID: dbID,

